### PR TITLE
add `no_std` support

### PR DESCRIPTION
**Pull request by zzhengzhuo**
add `no_std` support

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -71,6 +71,13 @@
 //! Guessing the proportion of ASCII vs. non-ASCII should be particularly
 //! feasible.
 
+#![no_std]
+
+#[cfg(feature = "alloc")]
+extern crate alloc;
+#[cfg(feature = "std")]
+extern crate std;
```
Comment (hsivonen) [inline src/lib.rs:79]: Why is this useful instead of always depending on `alloc`?

File: src/lib.rs [hunk 2]
```diff
@@ -71,6 +71,13 @@
 //! Guessing the proportion of ASCII vs. non-ASCII should be particularly
 //! feasible.
 
+#![no_std]
+
+#[cfg(feature = "alloc")]
+extern crate alloc;
```
Comment (hsivonen) [inline src/lib.rs:77]: What happens if you turn off default features and don't request either of `alloc` or `std`?

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,4 @@
 /target
 **/*.rs.bk
 Cargo.lock
+.history
```
Comment (hsivonen) [inline .gitignore]: No newline at the end of file.

File: src/lib.rs (hunk 1 above)
Comment (zzhengzhuo) [reply src/lib.rs:79]: emmm, maybe there are some apis that not depend on `alloc` or `std`, and can decide to compile with `serde/std` or `serde/no_std`.

File: src/lib.rs (hunk 2 above)
Comment (zzhengzhuo) [reply src/lib.rs:77]: Then you will not be able to use `std::vec` or `alloc::vec` and other alloction library.

File: .gitignore (hunk 3 above)
Comment (zzhengzhuo) [reply .gitignore]: ok

File: src/lib.rs (hunk 1 above)
Comment (zzhengzhuo) [reply src/lib.rs:79]: Will always depending on `alloc` be better? I cannot decide. 
`encoding-rs` seems always depend on `alloc`, maybe `charset` should depend on `alloc` by default when not depend on `std` and remove `alloc` feature.
What's your opinion?

**Comment by hsivonen** [comment]
0.1.4 now makes the crate `no_std` plus `alloc` in a simpler way.

**Pull request closed without merging**